segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/hdfs/HDFSStorageTest.java

```diff
@@ -67,6 +84,84 @@ public void tearDown() {
     }
 
     //region Fencing tests
+    /**
+     * A special case Fencing test that verifies the HDFSStorage fencing mechanism when the "fenced-out" instance keeps
+     * trying to write continuously. This verifies that any ongoing writes are properly handled upon fencing.
+     * By having the "fenced-out" instance keep writing, we will exercise the case when we have an ongoing write while
+     * a HDFS file is being set as read-only (the HDFS behavior in this case is that the ongoing write will complete).
+     */
+    @Test(timeout = 60000)
+    public void testZombieFencing() throws Exception {
+        final long epochCount = 30;
+        final int writeSize = 1000;
+        final String segmentName = "Segment";
+        val writtenData = new EnhancedByteArrayOutputStream();
+        final Random rnd = new Random(0);
+        int currentEpoch = 1;
+
+        // Create initial adapter.
+        val currentStorage = new AtomicReference<Storage>();
+        currentStorage.set(createStorage());
+        currentStorage.get().initialize(currentEpoch);
+
+        // Create the Segment and open it for the first time.
+        val currentHandle = new AtomicReference<SegmentHandle>(
+                currentStorage.get().create(segmentName, TIMEOUT)
+                              .thenCompose(v -> currentStorage.get().openWrite(segmentName))
+                              .get(TIMEOUT.toMillis(), TimeUnit.MILLISECONDS));
+
+        // Run a number of epochs.
+        while (currentEpoch <= epochCount) {
+            val oldStorage = currentStorage.get();
+            val handle = currentHandle.get();
+            val writeBuffer = new byte[writeSize];
+            val appends = Futures.loop(
+                    () -> true,
+                    () -> {
+                        rnd.nextBytes(writeBuffer);
+                        return oldStorage.write(handle, writtenData.size(), new ByteArrayInputStream(writeBuffer), writeBuffer.length, TIMEOUT)
+                                         .thenRun(() -> writtenData.write(writeBuffer));
+                    },
+                    executorService());
+
+            // Create a new Storage adapter with a new epoch and open-write the Segment, remembering its handle.
+            val newStorage = createStorage();
+            try {
+                newStorage.initialize(++currentEpoch);
+                currentHandle.set(newStorage.openWrite(segmentName).get(TIMEOUT.toMillis(), TimeUnit.MILLISECONDS));
+            } catch (Exception ex) {
+                newStorage.close();
+                throw ex;
+            }
+
+            currentStorage.set(newStorage);
+            try {
+                appends.get(TIMEOUT.toMillis(), TimeUnit.MILLISECONDS);
+                Assert.fail("Continuous appends on older epoch Adapter did not fail.");
```

Comment:
I think this assertion is here only as a safety guard because in principle we should never get here, even in the case the that the appends incorrectly keep going (it is an infinite loop after all).

Reply:
Ah, that's right, you can get here via the get timeout. Thanks for the clarification.